src/Redis/Orleans.Reminders.Redis/Storage/RedisReminderTable.cs

```diff
@@ -42,98 +44,147 @@ internal class RedisReminderTable : IReminderTable
-                RedisValue[] values2 = await _db.SortedSetRangeByValueAsync(RemindersRedisKey, "[\"00000000\",\"", to);
-                values = values1.Concat(values2);
+                throw new RedisRemindersException(Invariant($"{exception.GetType()}: {exception.Message}"));
             }
-
-            IEnumerable<ReminderEntry> records = values.Select(v => ConvertToEntry(v));
-            return new ReminderTableData(records);
         }
 
         public async Task<bool> RemoveRow(GrainId grainId, string reminderName, string eTag)
         {
-            (RedisValue from, RedisValue to) = GetFilter(grainId, reminderName, eTag);
-            long removed = await _db.SortedSetRemoveRangeByValueAsync(RemindersRedisKey, from, to);
-            return removed > 0;
+            try
+            {
+                (RedisValue from, RedisValue to) = GetFilter(grainId, reminderName, eTag);
+                long removed = await _db.SortedSetRemoveRangeByValueAsync(_hashSetKey, from, to);
+                return removed > 0;
+            }
+            catch (Exception exception)
+            {
+                throw new RedisRemindersException(Invariant($"{exception.GetType()}: {exception.Message}"));
+            }
         }
 
         public async Task TestOnlyClearTable()
         {
-            await _db.ExecuteAsync("FLUSHDB");
+            try
+            {
+                await _db.KeyDeleteAsync(_hashSetKey);
+            }
+            catch (Exception exception)
+            {
+                throw new RedisRemindersException(Invariant($"{exception.GetType()}: {exception.Message}"));
+            }
         }
 
         public async Task<string> UpsertRow(ReminderEntry entry)
         {
-            if (_logger.IsEnabled(LogLevel.Debug))
-            {
-                _logger.LogDebug("UpsertRow entry = {Entry}, ETag = {ETag}", entry.ToString(), entry.ETag);
-            }
-
-            (string etag, string value) = ConvertFromEntry(entry);
-            (string from, string to) = GetFilter(entry.GrainId, entry.ReminderName);
-
-            ITransaction tx = _db.CreateTransaction();
-            _db.SortedSetRemoveRangeByValueAsync(RemindersRedisKey, from, to).Ignore();
-            _db.SortedSetAddAsync(RemindersRedisKey, value, 0).Ignore();
-            bool success = await tx.ExecuteAsync();
-            if (success)
+            try
             {
-                return etag;
+                if (_logger.IsEnabled(LogLevel.Debug))
+                {
+                    _logger.LogDebug("UpsertRow entry = {Entry}, ETag = {ETag}", entry.ToString(), entry.ETag);
+                }
+
+                (string etag, string value) = ConvertFromEntry(entry);
+                (string from, string to) = GetFilter(entry.GrainId, entry.ReminderName);
+
+                ITransaction tx = _db.CreateTransaction();
+                _db.SortedSetRemoveRangeByValueAsync(_hashSetKey, from, to).Ignore();
```

Comment:
if we're updating a single item, are we sure we want to do this by score? `zrem` is more efficient than `zremrangebyscore`, and can *probably* just do as a single `zadd` which (unless `incr` is specified) effectively does a re-insert with the new score *anyway*

Reply:
I don't quite understand this comment, sorry. What are you proposing in terms of API calls?

Reply:
let me just check my understanding of the setup here first; so a redis sorted set is an order set of token + score pairs (ordered by score); here, the `SortedSetRemoveRangeByValueAsync` maps to `zremrangebylex`, which removes a range *by key* (which is the non-performant axis - i.e. the non-sorted axis) - but it *looks* like we're really just updating a single entity here? now; I can't tell without squinting really hard, but: is the `value` here changing? if not, we can *just* issue the `_db.SortedSetAddAsync(_hashSetKey, value, 0)`, which will effectively just:

- locate the item with key `value` if it exists, and change the score to `0`, effectively moving the item in the sorted set
- if it doesn't exist, just add `value` with score `0`

If I've misunderstood what we're storing here and the data we are removing *isn't* keyed with `value`, then: fine - my bad

---

unrelated; I don't know what `.Ignore()` does, but whatever it is: you can probably just use:

``` c#
_ = _db.SortedSetRemoveRangeByValueAsync(_hashSetKey, from, to, flags: CommandFlags.FireAndForget);
```

which will cut out a few bits of internal machinery (and never fault)

Reply:
> which is the non-performant axis - i.e. the non-sorted axis
I believe you, but what is your source for this? The Redis docs state the time complexity for both as being `O(Log(N)+M)`

Reply:
Ideally, what we want is:
* To be able to perform range queries on a key value (hash code of grain id + grain id)
  * Currently, we are using a lexicographic sort with the format `[HashCode,GrainId,ReminderName,ETag,...other elements...]`
* To be able to ensure that if an ETag is provided, that the operation only proceeds if any existing entry matches that ETag (it is part of the key value)

Is there a way that we can add a transaction condition that there must be at least one value in the range `[from,to]`? I am hoping we don't need to resort to Lua, but perhaps we do.

Reply:
I think we should use a Lua script instead of a transaction...

Reply:
Lua is almost always both simpler and more efficient than the WATCH/MULTI/EXEC beast; I support this notion